### PR TITLE
Add logic gates to generators

### PR DESCRIPTION
**Pull request by arcondello**
Closes https://github.com/dwavesystems/dimod/issues/951

### CODE REVIEW

File: dimod/generators/gates.py [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+    bqm.add_variable(in0)
+    bqm.add_variable(in1)
+    bqm.add_variable(out, bias=3)
+
+    # add the quadratic biases
+    bqm.add_quadratic(in0, in1, 1)
+    bqm.add_quadratic(in0, out, -2)
+    bqm.add_quadratic(in1, out, -2)
+
+    # the bqm currently as a strength of 1, so just need to scale
```
Comment (hhtong) [inline dimod/generators/gates.py:57]: typo: has (same for other generators below)

**Review by hhtong — APPROVED** [review]
LGTM!

File: dimod/generators/gates.py [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+def and_gate(in0: Variable, in1: Variable, out: Variable,
+             *, strength: float = 1.0) -> BinaryQuadraticModel:
+    """Return a binary quadratic model with ground states corresponding to an
+    AND gate.
+
+    Args:
+        in0: The variable label for one of the inputs.
+        in1: The variable label for one of the inputs.
+        out: The variable label for the output.
+        strength: The energy of the first infeasible state.
```
Comment (JoelPasvolsky) [inline dimod/generators/gates.py:38]: "first" --> we talk about the "first excited state" and that makes sense but here first is likely to be understood as the first combination of the 8 `xxx` binary states rather then the lowest-energy infeasible state.

File: dimod/generators/gates.py [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+    Args:
+        in0: The variable label for one of the inputs.
+        in1: The variable label for one of the inputs.
+        in2: The variable label for one of the inputs
+        sum_: The variable label for the sum output.
+        carry: The variable label for the carry output.
+        strength: The energy of the first infeasible state.
+
+    Returns:
+        A binary quadratic model with ground states corresponding to a half
```
Comment (JoelPasvolsky) [inline dimod/generators/gates.py:79]: "half" --> this is the full-adder gate

**Pull request merged**